cli/flox/src/commands/upgrade.rs

```diff
@@ -41,51 +48,319 @@ impl Upgrade {
         );
         let result = span.in_scope(|| {
-            environment.upgrade(
-                &flox,
-                &self
-                    .groups_or_iids
-                    .iter()
-                    .map(String::as_str)
-                    .collect::<Vec<_>>(),
-            )
+            let groups_or_iids = &self
+                .groups_or_iids
+                .iter()
+                .map(String::as_str)
+                .collect::<Vec<_>>();
+
+            if self.dry_run {
+                environment.dry_upgrade(&flox, groups_or_iids)
+            } else {
+                environment.upgrade(&flox, groups_or_iids)
+            }
         })?;
 
-        let upgraded = result.packages().collect::<Vec<_>>();
+        let diff = result.diff_for_system(&flox.system);
 
-        if upgraded.is_empty() {
+        if diff.is_empty() {
             if self.groups_or_iids.is_empty() {
                 message::plain(format!(
-                    "ℹ️  No packages need to be upgraded in environment {description}."
+                    "No package(s) need to be upgraded in {description}."
                 ));
             } else {
                 message::plain(format!(
-                    "ℹ️  The specified packages do not need to be upgraded in environment {description}."
-                 ) );
-            }
-        } else {
-            for package in upgraded {
-                message::plain(format!(
-                    "⬆️  Upgraded '{package}' in environment {description}."
+                    "The specified package(s) do not need to be upgraded in {description}."
                 ));
             }
+            return Ok(());
+        }
+
+        let rendered_diff = render_diff(&diff);
+        let num_changes = diff.len();
+
+        if self.dry_run {
+            message::plain(formatdoc! {"
+                Dry run: Upgrade will apply build changes to {num_changes} package(s) in {description}:
+                {rendered_diff}
+
+                To apply these changes, run the command without the '--dry-run' flag.
+            "});
 
-            warn_manifest_changes_for_services(&flox, environment.as_ref());
+            return Ok(());
         }
 
+        message::plain(formatdoc! {"
+            ✅  Upgrade sucessfully applied build changes to {num_changes} package(s) in {description}:
+            {rendered_diff}
+        "});
+
+        warn_manifest_changes_for_services(&flox, environment.as_ref());
+
         Ok(())
     }
 }
+
+/// Render a diff of locked packages before and after an upgrade
+fn render_diff(diff: &[(LockedPackage, LockedPackage)]) -> String {
+    diff.iter()
+        .map(|(before, after)| {
+            let install_id = before.install_id();
+            let old_version = before.version().unwrap_or("unknown");
+            let new_version = after.version().unwrap_or("unknown");
+
+            if new_version == old_version {
+                format!("- {install_id}: {old_version}")
```

Comment:
```suggestion
                format!("- {install_id}: dependencies upgraded")
```

Reply:
Dependencies does not quite cover possible changes to the drv, hence I had "build changes" in this position first but we decided to move that into the general success description.